### PR TITLE
Adding bootstrap mode for Pinot-hadoop job to output segments into relative directories.

### PR DESCRIPTION
**Pull request by fx19880617**
1. Skip hidden files or temp files created by computation frameworks like hadoop, spark.
2. Adding a `job.bootstrap` flag to make output directory following the relative paths from input path.

**sample-job.properties**
```
input.dir = /path/to/input
output.dir = /path/to/output
job.bootstrap=true
segment.table.name=mytable
```
Assume data structure under `/path/to/input` is like:
```
/path/to/input/yyyy=2019/mm=10/dd=1/part-0-r-aaa.avro
/path/to/input/yyyy=2019/mm=10/dd=2/part-0-r-bbb.avro
/path/to/input/yyyy=2019/mm=10/dd=3/part-0-r-ccc.avro
```

We expect the output directory structure to be:
```
/path/to/output/yyyy=2019/mm=10/dd=1/mytable_0.tar.gz
/path/to/output/yyyy=2019/mm=10/dd=2/mytable_1.tar.gz
/path/to/output/yyyy=2019/mm=10/dd=3/mytable_2.tar.gz
```

In the old job, we will get:
```
/path/to/output/mytable_0.tar.gz
/path/to/output/mytable_1.tar.gz
/path/to/output/mytable_2.tar.gz
```

### CODE REVIEW

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java [hunk 1]
```diff
@@ -259,6 +271,24 @@ protected void map(LongWritable key, Text value, Context context)
...
+   *    inputFile = "/path/to/input/a/b/c/d.avro"
+   *    outputDir = "/path/to/output"
+   *    getBootstrapOutputPath(baseInputDir, inputFile, outputDir) = /path/to/output/a/b/c
+   */
+  protected static Path getBootstrapOutputPath(URI baseInputDir, URI inputFile, Path outputDir) {
+    URI relativePath = baseInputDir.relativize(inputFile);
+    if (relativePath.getPath().length() > 0) {
+      return new Path(outputDir, relativePath.getPath()).getParent();
+    }
+    return null;
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: In what scenario can this return null? Can it happen in normal case? Should throw exception instead IMO

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/JobConfigConstants.java [hunk 2]
```diff
@@ -59,4 +59,5 @@
   // This setting should be used if you will generate less # of segments after
   // push. In preprocessing, this is likely because we resize segments.
   public static final String DELETE_EXTRA_SEGMENTS = "delete.extra.segments";
+  public static final String BOOTSTRAP_JOB = "job.bootstrap";
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/JobConfigConstants.java]: Recommend using something more specific such as "use.relative.path". I don't see a clear connection between bootstrap and using relative path.

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java (hunk 1 above)
Comment (fx19880617) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: good point. Change this to throw RTE

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/JobConfigConstants.java (hunk 2 above)
Comment (fx19880617) [reply pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/JobConfigConstants.java]: done

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java [hunk 3]
```diff
@@ -66,6 +67,7 @@
   protected String _rawTableName;
   protected Schema _schema;
   protected SegmentNameGenerator _segmentNameGenerator;
+  protected boolean _bootstrapJob = false;
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: Change this variable to `_useRelativePath`

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java [hunk 4]
```diff
@@ -259,6 +270,24 @@ protected void map(LongWritable key, Text value, Context context)
...
+   * Generate an output directory path for bootstrap mode.
+   * This method will compute the relative path based on `inputFile` and `baseInputDir`,
+   * then apply only the directory part of relative path to `outputDir`.
+   * E.g.
+   *    baseInputDir = "/path/to/input"
+   *    inputFile = "/path/to/input/a/b/c/d.avro"
+   *    outputDir = "/path/to/output"
+   *    getBootstrapOutputPath(baseInputDir, inputFile, outputDir) = /path/to/output/a/b/c
+   */
+  protected static Path getBootstrapOutputPath(URI baseInputDir, URI inputFile, Path outputDir) {
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: `getRelativeOutputPath()`, and adjust the javadoc accordingly

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java [hunk 5]
```diff
@@ -259,6 +270,24 @@ protected void map(LongWritable key, Text value, Context context)
...
+    URI relativePath = baseInputDir.relativize(inputFile);
+    if (relativePath.getPath().length() > 0 && !relativePath.equals(inputFile)) {
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: Change this to a Preconditions.checkState() for readability?

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java [hunk 6]
```diff
@@ -251,6 +254,14 @@ protected void map(LongWritable key, Text value, Context context)
         DataSize.fromBytes(uncompressedSegmentSize), DataSize.fromBytes(compressedSegmentSize));
 
     Path hdfsSegmentTarFile = new Path(_hdfsSegmentTarDir, segmentTarFileName);
+    if (_bootstrapJob) {
+      Path bootstrapOutputPath =
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: Rename this variable to relativeOutputPath

File: pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java [hunk 7]
```diff
@@ -251,6 +254,14 @@ protected void map(LongWritable key, Text value, Context context)
...
+          getBootstrapOutputPath(new Path(_jobConf.get(JobConfigConstants.PATH_TO_INPUT)).toUri(),
+              hdfsInputFile.toUri(), _hdfsSegmentTarDir);
+      if (bootstrapOutputPath != null) {
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationMapper.java]: You can remove this check

File: pinot-hadoop/src/test/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+
+public class SegmentCreationTest {
+
+  @Test
+  public void testBootstrapOutputPath() {
+    URI baseDir = new Path("/path/to/input").toUri();
+    URI inputFile = new Path("/path/to/input/a/b/c/d/e").toUri();
+    Path outputPath = new Path("/path/to/output");
+    Path bootstrapOutputPath = SegmentCreationMapper.getBootstrapOutputPath(baseDir, inputFile, outputPath);
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/test/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationTest.java]: Can you change all the bootstrapOutputPath to relativeOutputPath in the tests?

File: pinot-hadoop/src/test/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationTest.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    Assert.assertEquals(bootstrapOutputPath.toString(), "/path/to/output/a/b/c/d");
+  }
+
+
+  @Test
+  public void testBootstrapOutputPath1() {
+    URI baseDir = new Path("/path/to/input/*/*.avro").toUri();
+    URI inputFile = new Path("/path/to/input/a/b.avro").toUri();
+    Path outputPath = new Path("/path/to/output");
+    Assert.assertThrows(() -> SegmentCreationMapper.getBootstrapOutputPath(baseDir, inputFile, outputPath));
```
Comment (Jackie-Jiang) [inline pinot-hadoop/src/test/java/org/apache/pinot/hadoop/job/mappers/SegmentCreationTest.java]: Cool. Don't know we can directly assert throws haha

**Pull request merged**